index.js

```diff
@@ -0,0 +1,254 @@
+            if (count >= minChunks) {
+                commonModules.add(module);
+                globalModulesSet.add(module);
+            }
+        }
+
+        const {moduleDeps, chunkDeps} =
+            this.getDependencies(commonModules, globalModulesSet);
+
+        // If there aren't any modules to add to the chunk don't bother
+        // creating a shared chunk.
+        if (commonModules.size === 0 && moduleDeps.size === 0) {
+            console.warn(
+                `'${spec.name}' chunk not created, contains no modules`);
+            return;
+        }
+
+        // The target chunk is the new shared chunk to which we will be adding
+        // modules that are used by one or more chunks.
+        const targetChunk = allChunksNameMap.has(spec.name)
+        ? allChunksNameMap.get(spec.name)
+        : compilation.addChunk(spec.name);
+
+        // Add module dependencies.
+        for (const dep of moduleDeps) {
+            commonModules.add(dep);
+            globalModulesSet.add(dep);
+        }
+
+        // Add all common modules to the target chunk.
+        for (const module of commonModules) {
+            targetChunk.addModule(module);
+            module.addChunk(targetChunk);  // modules can appear in multiple chunks
+        }
+
+        // Add chunk dependencies.
+        targetChunk.parents = [...chunkDeps];
+        for (const chunk of chunkDeps) {
+            chunk.addChunk(targetChunk);
+        }
+
+        // It's possible that moduleFilter function filters out all modules
+        // from a particular chunk in which case it isn't actually affected.
+        // TODO(kevinb): figure out what to do when extracting common
+        // chunks from async chunks, see CommonsChunkPlugin
+        const actuallyAffectedChunks = new Set();
+        for (const module of commonModules) {
+            for (const chunk of affectedChunks) {
+                if (module.removeChunk(chunk)) {
+                    actuallyAffectedChunks.add(chunk);
+                }
+            }
+        }
+
+        // Copied from makeTargetChunkParentOfAffectedChunks in
+        // CommonsChunkPlugin.js
+        for (const chunk of actuallyAffectedChunks) {
+            chunk.parents = [...chunk.parents, targetChunk];
+            targetChunk.addChunk(chunk);
+
+            for (const entrypoint of chunk.entrypoints) {
+                entrypoint.insertChunk(targetChunk, chunk);
+            }
+        }
+    }
+
+    // Return module and chunk dependencies for the given commonModules.
+    // Module dependencies are determined recursively.  Whenever a module
+    // appearing in globalModulesSet is encountered, recursion is terminated
+    // and any chunks that the module appears in are return as chunk
+    // dependencies.
+    getDependencies(commonModules, globalModulesSet) {
+        const moduleDeps = new Set();
+        const chunkDeps = new Set();
+
+        const walkDeps = (module) => {
```

Comment:
It seems like this can be made a static global function, as well (taking a module and a globalModuleSet).